portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -107,5 +123,43 @@ private Integer getMaxComments(HttpServletRequest request) {
     }
     return numMaxComments;
   }
+
+  //method from source code
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's dev server, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
```

Comment:
Hmm this will always return the path only, assuming the input is a valid URL (which it probably always is?). If the value of `getServingUrl` is on a different domain vs. your website, this probably would not work when deployed.

Is there a way to do a more explicit check, something like

```
if (isDevServer()) {
  return new URL(imagesService.getServingUrl(options)).getPath();
} else {
  return imagesService.getServingUrl(options);
}
```

(I'm not sure how isDevServer would be implemented; it looks like there are instructions for Python [here](https://cloud.google.com/appengine/docs/standard/python/tools/using-local-server#detecting_application_runtime_environment) though which involve checking an environment variable, Java might work the same way)

Reply:
I'll look into this and test it out! This will probably be helpful with the current deployment issues.

Reply:
It looks like there may be problems with the getServingUrl method upon deployment, so I went ahead and removed that and instead stored the blobKey directly into datastore. These changes are reflected in the commits below.